modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/resolver/MetadataContextArgumentResolver.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.runtime.resolver;
+
+import static org.mule.module.extension.internal.ExtensionProperties.ENCODING_PARAMETER_NAME;
+import static org.mule.module.extension.internal.ExtensionProperties.METADATA_CONTEXT;
+import static org.mule.module.extension.internal.ExtensionProperties.MIME_TYPE_PARAMETER_NAME;
+import org.mule.api.MuleMessage;
+import org.mule.extension.api.runtime.ContentType;
+import org.mule.extension.api.runtime.FixedContextMetadata;
+import org.mule.extension.api.runtime.ContentMetadata;
+import org.mule.extension.api.runtime.MutableContentMetadata;
+import org.mule.extension.api.runtime.OperationContext;
+import org.mule.module.extension.internal.ExtensionProperties;
+import org.mule.module.extension.internal.runtime.OperationContextAdapter;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Returns the value of the {@link ExtensionProperties#METADATA_CONTEXT} variable,
+ * which is expected to have been previously set on the supplied {@link OperationContext}.
+ * <p>
+ * Notice that for this to work, the {@link OperationContext}
+ * has to be an instance of {@link OperationContextAdapter}
+ *
+ * @since 4.0
+ */
+public final class MetadataContextArgumentResolver implements ArgumentResolver<ContentMetadata>
+{
+
+    /**
+     * @param operationContext an {@link OperationContext}
+     * @return the value of {@link OperationContextAdapter#getVariable(String)} using the {@link ExtensionProperties#METADATA_CONTEXT}
+     * @throws ClassCastException       if {@code operationContext} is not an instance of {@link OperationContextAdapter}
+     * @throws IllegalArgumentException if the variable has not been set on the {@code operationContext}
+     */
+    @Override
+    public ContentMetadata resolve(OperationContext operationContext)
+    {
+        final OperationContextAdapter operationContextAdapter = (OperationContextAdapter) operationContext;
```

Comment:
This cast is all over the place and we'll have to live with it until the mule-api is available. Then I'll be my genuine pleasure to go kill it. (MULE-8946)